acbuild/script.go

```diff
@@ -0,0 +1,188 @@
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	nestedScriptEnvVar = "ACBUILD_NESTED_SCRIPT"
+)
+
+var (
+	cmdScript = &cobra.Command{
+		Use:     "script SCRIPT_FILE",
+		Short:   "Runs an acbuild script",
+		Example: "acbuild script build-myapp.acb",
+		Run:     runWrapper(runScript),
+	}
+)
+
+func init() {
+	cmdAcbuild.AddCommand(cmdScript)
+}
+
+func runScript(cmd *cobra.Command, args []string) (exit int) {
+	if len(args) != 1 {
+		cmd.Usage()
+		return 1
+	}
+
+	scriptName := args[0]
+	rawScript, err := ioutil.ReadFile(args[0])
+	if err != nil {
+		stderr("script: %v", err)
+		return getErrorCode(err)
+	}
+
+	if debug {
+		stderr("Running script from %s", scriptName)
+	}
+
+	err = execScript(rawScript)
+	if err != nil {
+		stderr("script: %v", err)
+		return getErrorCode(err)
+	}
+	return 0
+}
+
+func execScript(rawScript []byte) error {
+	script := strings.Split(string(rawScript), "\n")
+	for i := 0; i < len(script); i++ {
+		script[i] = trimComments(script[i])
+
+		if strings.HasPrefix(script[i], "run") && os.Geteuid() != 0 {
+			return fmt.Errorf("scripts using the run subcommand must be run as root")
+		}
+	}
+
+	var tmpDir string
+	nestedScript := false
+	if inheritedPath := os.ExpandEnv("$" + nestedScriptEnvVar); inheritedPath != "" {
+		fmt.Printf("nested script! %s\n", inheritedPath)
+		tmpDir = inheritedPath
+		nestedScript = true
+	} else {
+		var err error
+		tmpDir, err = ioutil.TempDir("", "acbuild")
+		if err != nil {
+			return err
+		}
+		defer os.RemoveAll(tmpDir)
+		contextpath = tmpDir
+	}
+
+	for _, line := range script {
+		if line != "" {
+			err := execACBuild(tmpDir, line)
+			if err != nil {
+				if !strings.HasPrefix(line, "begin") && !nestedScript {
+					err1 := newACBuild().End()
+					if err1 != nil {
+						stderr("script: %v", err1)
+					}
+				}
+				return err
+			}
+		}
+	}
+	if !nestedScript {
+		err := newACBuild().End()
+		if err != nil {
+			return err
+		}
+	}
+	if debug {
+		stderr("Script has been completed")
+	}
+
+	return nil
+}
+
+func execACBuild(workPath, line string) error {
+	args := []string{"--debug", "--work-path=" + workPath}
+	args = append(args, tokenizeLine(line)...)
+	cmd := exec.Command(os.Args[0], args...)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Env = os.Environ()
+	if strings.HasPrefix(line, "sc") {
+		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", nestedScriptEnvVar, workPath))
+	}
+	return cmd.Run()
+}
+
+func trimComments(line string) string {
+	trimAt := -1
+	isEscaped := false
+	for i, char := range line {
+		if isEscaped {
+			isEscaped = false
+			continue
+		}
+		switch char {
+		case '\\':
+			isEscaped = true
+		case '#':
+			trimAt = i
+			break
+		}
+	}
+	if trimAt != -1 {
+		line = line[:trimAt]
+	}
+	return strings.TrimSpace(line)
+}
+
+func tokenizeLine(line string) []string {
```

Comment:
I'd like to see some nice table testing for this method. Tokenizer bugs are annoying.